base/cvd/cuttlefish/host/commands/cvd/instances/instance_record.cpp

```diff
@@ -83,4 +96,94 @@ Result<Json::Value> LocalInstance::FetchStatus(std::chrono::seconds timeout) {
   return CF_EXPECT(FetchInstanceStatus(*this, timeout));
 }
 
+Result<void> LocalInstance::PressPowerBtn() {
+  Command cmd(
+      CF_EXPECT(HostToolTarget(host_artifacts_path()).GetPowerBtnBinPath()));
+
+  cmd.AddParameter("--instance_num=", id());
+  cmd.SetEnvironment({});
+  AddEnvironment(cmd, *this);
+
+  LOG(DEBUG) << "Executing: " << cmd.ToString();
+
+  siginfo_t infop;
+  cmd.Start().Wait(&infop, WEXITED);
+  CF_EXPECT(CheckProcessExitedNormally(infop));
+
+  return {};
+}
+
+Result<void> LocalInstance::Restart(std::chrono::seconds launcher_timeout,
+                                    std::chrono::seconds boot_timeout) {
+  SharedFD monitor_socket = CF_EXPECT(GetLauncherMonitor(launcher_timeout));
+
+  LOG(INFO) << "Requesting restart";
+  CF_EXPECT(RunLauncherAction(monitor_socket, LauncherAction::kRestart,
+                              launcher_timeout.count()));
+
+  LOG(INFO) << "Waiting for device to boot up again";
+  CF_EXPECT(WaitForRead(monitor_socket, boot_timeout.count()));
+  RunnerExitCodes boot_exit_code = CF_EXPECT(ReadExitCode(monitor_socket));
+  CF_EXPECT(boot_exit_code != RunnerExitCodes::kVirtualDeviceBootFailed,
+            "Boot failed");
+  CF_EXPECT(boot_exit_code == RunnerExitCodes::kSuccess,
+            "Unknown response" << static_cast<int>(boot_exit_code));
+
+  LOG(INFO) << "Restart successful";
+  return {};
+}
```

Comment:
This makes the launcher control socket the API boundary between `cuttlefish-base` and `device/google/cuttlefish`, rather than executables.

The immediate problem I see there is that this sets the split between `LauncherAction` and `ExtendedLauncherAction` in stone. Does it seem valuable to finish the migration to `ExtendedLauncherAction` first to make it possible to give fields to the existing actions, or is it fine to keep supporting both kinds of `LauncherAction` indefinitely?

If we go through with this, it seems worth documenting the dependency on the AOSP side.